### ISSUE TITLE
async restore will fail with "unable to get WAL file" if /tmp/pgbackrest is not writeable 

### ISSUE THREAD

**Issue by geirra** [pgbackrest/pgbackrest#1722]
1. pgBackRest version:
```
2.38-1.rhel8
```
2. PostgreSQL version:
``` 
12.10
```

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
```
RHEL 8.5
```

4. Did you install pgBackRest from source or from a package?
```
Package
```

5. Please attach the following as applicable:
```
[global]
repo1-path=/mnt/restore/path
repo1-retention-full=10
repo1-retention-full-type=time
log-level-console=debug
log-level-file=debug
spool-path=/var/lib/pgsql/12/data/pgbackrest
archive-async=y
compress-type=zst
process-max=6

[global:archive-get]
process-max=4

[global:archive-push]
process-max=6

[my-stanza]
pg1-path=/var/lib/pgsql/12/data
```

Detailed log
```
2022-04-22 15:25:48.725 CEST [2860677] LOG:  database system was interrupted; last known up at 2022-04-22 07:08:20 CEST
2022-04-22 15:25:48.828 P00   INFO: archive-get command begin 2.38: [00000002.history, pg_wal/RECOVERYHISTORY] --archive-async --exec-id=2860678-5660aae8 --log-level-console=detail --log-level-file=detail --pg1-path=/var/lib/pgsql/12/data --process-max=4 --repo1-path=/mnt/restore/path --spool-path=/var/lib/pgsql/12/data/pgbackrest --stanza=my-stanza
2022-04-22 15:25:48.848 P00   INFO: unable to find 00000002.history in the archive
2022-04-22 15:25:48.848 P00   INFO: archive-get command end: completed successfully (21ms)
2022-04-22 15:25:48.849 CEST [2860677] LOG:  starting archive recovery
2022-04-22 15:25:48.855 P00   INFO: archive-get command begin 2.38: [00000001000377A80000007C, pg_wal/RECOVERYXLOG] --archive-async --exec-id=2860679-118f8a90 --log-level-console=detail --log-level-file=detail --pg1-path=/var/lib/pgsql/12/data --process-max=4 --repo1-path=/mnt/restore/path --spool-path=/var/lib/pgsql/12/data/pgbackrest --stanza=my-stanza
2022-04-22 15:25:55.354 CEST [2860696] FATAL:  the database system is starting up
2022-04-22 15:26:07.856 CEST [2860730] FATAL:  the database system is starting up
2022-04-22 15:26:07.904 CEST [2860731] FATAL:  the database system is starting up
2022-04-22 15:26:13.047 CEST [2860929] FATAL:  the database system is starting up
2022-04-22 15:26:16.783 CEST [2860943] FATAL:  the database system is starting up
2022-04-22 15:26:17.995 CEST [2860947] FATAL:  the database system is starting up
2022-04-22 15:26:27.135 CEST [2860970] FATAL:  the database system is starting up
2022-04-22 15:26:28.259 CEST [2860972] FATAL:  the database system is starting up
2022-04-22 15:26:29.309 CEST [2860975] FATAL:  the database system is starting up
2022-04-22 15:26:30.346 CEST [2860979] FATAL:  the database system is starting up
2022-04-22 15:26:30.355 CEST [2860984] FATAL:  the database system is starting up
2022-04-22 15:26:30.362 CEST [2860986] FATAL:  the database system is starting up
2022-04-22 15:26:33.551 CEST [2861009] FATAL:  the database system is starting up
2022-04-22 15:26:34.674 CEST [2861013] FATAL:  the database system is starting up
2022-04-22 15:26:37.785 CEST [2861016] FATAL:  the database system is starting up
2022-04-22 15:26:37.896 CEST [2861017] FATAL:  the database system is starting up
2022-04-22 15:26:39.201 CEST [2861020] FATAL:  the database system is starting up
2022-04-22 15:26:41.840 CEST [2861024] FATAL:  the database system is starting up
2022-04-22 15:26:44.179 CEST [2861033] FATAL:  the database system is starting up
2022-04-22 15:26:46.242 CEST [2861043] FATAL:  the database system is starting up
ERROR: [082]: unable to get WAL file '00000001000377A80000007C' from the archive asynchronously after 60 second(s)
2022-04-22 15:26:48.878 P00   INFO: archive-get command end: aborted with exception [082]
2022-04-22 15:26:48.878 CEST [2860677] LOG:  invalid checkpoint record
2022-04-22 15:26:48.878 CEST [2860677] FATAL:  could not locate required checkpoint record
2022-04-22 15:26:48.878 CEST [2860677] HINT:  If you are restoring from a backup, touch "/var/lib/pgsql/12/data/recovery.signal" and add required recovery options.
	If you are not restoring from a backup, try removing the file "/var/lib/pgsql/12/data/backup_label".
	Be careful: removing "/var/lib/pgsql/12/data/backup_label" will result in a corrupt cluster if restoring from a backup.
2022-04-22 15:26:48.879 CEST [2860675] LOG:  startup process (PID 2860677) exited with exit code 1
2022-04-22 15:26:48.879 CEST [2860675] LOG:  aborting startup due to startup process failure
2022-04-22 15:26:48.940 CEST [2860675] LOG:  database system is shut down
```

7. 
This one might be related to #1351 

During a restore test i had problems with rebuilding the database, the initial restore was fine.
But the rebuild based on the WAL files failed.

The log complained about the WAL file was missing, but the missing file was located in the archive for the wal files.

The issue was bypassed with changing 
```archive-async``` to ```off```  and the rebuild was successful.

Did some more testing and noticed that /tmp/pgbackrest was owned by root, changing permission on the folder so postgres could write here solved the problem while ```archive-async``` was set to```on``` 

I have logs with debug level for both these scenarios if needed
1. async on with root ownership of /tmp/pgbackrest
2. async on with correct permissions on /tmp/pgbackrest

I have initially not included these log files due to their size, if any data is needed i can provide them.

TLDR: 
async archive-get tries to use tmp folder, when pgbackrest doesn't find it files there it fails with WAL file missing.
There is no hints that this due the temp dir, from the logs it looks like the WAL archive is at fault.

**Comment by dwsteele** [comment]
There should have been an error in the archive-push-async log (can you verify?), but it seems like some hints would be useful here. Maybe something like:
```
ERROR: [082]: unable to get WAL file '00000001000377A80000007C' from the archive asynchronously after 60 second(s)
HINT: is '/tmp/pgbackrest' writable by the 'pgbackrest' user?
HINT: are there any errors in `/var/log/pgbackrest/my-stanza-archive-push-async.log`?
```
Another thing would be to make the foreground process error on any lock failure other than another process holding the lock. That would make it a lot more obvious what is going on.

**Comment by dwsteele** [comment]
Yeah, I think here (https://github.com/pgbackrest/pgbackrest/blob/126fc99c77998cc65094d92e955347100aa42531/src/common/lock.c#L324) it should only suppress the error when it is `EWOULDBLOCK`. That way permissions errors and such would be logged by the `archive_command`.

We would not need a hit for the lock file anymore (since there would be an error) so:
```
ERROR: [082]: unable to get WAL file '00000001000377A80000007C' from the archive asynchronously after 60 second(s)
HINT: are there any errors in `/var/log/pgbackrest/my-stanza-archive-push-async.log`?
```

**Comment by geirra** [comment]
In this case we simulated a DR scenario, where recovery had to be done on a new server. 
So there would be no push logs here, since this server is new. 

But the STANZA-archive-get-async.log is empty in this case.

**Comment by dwsteele** [comment]
> So there would be no push logs here, since this server is new.

Whoops. But get and push have the same behavior here so both would change.

> But the STANZA-archive-get-async.log is empty in this case.

Interesting. The log file should be opened before the lock is attempted. There must be something else going on, but at least exposing the lock error in `archive_command` is a good place to start.

**Comment by dwsteele** [comment]
Closing since PR #1730 has been created to fix this. Please have a look.

@geirra How would you like your name to appear in the release notes?

**Comment by dwsteele** [comment]
> About the name, shure why not :)

Not sure that's an answer. :) Would you prefer `geirra` or something else?

> Should i try the patch in our scenario test?

That would be fantastic. I think this will do what you need but additional testing is always a good thing.

**Comment by geirra** [comment]
Hi

Sorry for the late reply, i can confirm that the patch works and provides a detailed error :)
Great work  and thanks ! @dwsteele 

```
ERROR: [050]: unable to acquire lock on file '/tmp/pgbackrest/zabbix-archive.lock': Permission denied
       HINT: does the user running pgBackRest have permissions on the '/tmp/pgbackrest/zabbix-archive.lock' file?
```

About the name
```
Geir Råness
```

**Comment by dwsteele** [comment]
Excellent! Thank you for having a look at this.

> Sorry for the late reply

Not late at all. We don't expect you to work overnight... :)
